src/loaders/svg/tvgSvgLoader.cpp

```diff
@@ -1738,6 +1738,10 @@ static bool _attrParsePolygonPoints(const char* str, SvgPolygonNode* polygon)
 {
     float num;
     while (_parseNumber(&str, nullptr, &num)) polygon->pts.push(num);
+    if (polygon->pts.count % 2 != 0) {
```

Comment:
from https://www.w3schools.com/graphics/svg_polygon.asp

This polygon has 3 points
```
<svg height="220" width="500" xmlns="http://www.w3.org/2000/svg">
  <polygon points="100,10 150,190 50,190" style="fill:lime;stroke:purple;stroke-width:3" />
</svg>
```
while this polygon has 4 points
```
<svg height="260" width="500" xmlns="http://www.w3.org/2000/svg">
  <polygon points="220,10 300,210 170,250 123,234" style="fill:lime;stroke:purple;stroke-width:3" />
</svg>
```
Is it correct?

Reply:
> brokenFiles/default_1635658206_IDX_0_RAND_14009677174565202864.png

oh... the example I mentioned is wrong. Please ignore this.
These types of examples are the problem. Points contain uninterpretable letters or numbers.
`default_1635658206_IDX_156_RAND_537143208668833494.svg`

and

The `odds` I mentioned was about the size of the points array.
Heap-buffer-overflow occurs in the below case .
```
<svg height="260" width="500" xmlns="http://www.w3.org/2000/svg">
  <polygon points="220, 10     300,210     170,250     123" style="fill:lime;stroke:purple;stroke-width:3" />
</svg>
```
```
 $ ./build/examples/PictureSvg
tvg engine: software
=================================================================
==193713==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6030000d631c at pc 0x7f02df965279 bp 0x7ffed0620cb0 sp 0x7ffed0620ca0
READ of size 4 at 0x6030000d631c thread T0
    #0 0x7f02df965278 in _recognizeShape ../src/loaders/svg/tvgSvgSceneBuilder.cpp:415
    #1 0x7f02df96ff70 in _appendShape ../src/loaders/svg/tvgSvgSceneBuilder.cpp:452
    #2 0x7f02df96ff70 in _shapeBuildHelper ../src/loaders/svg/tvgSvgSceneBuilder.cpp:389
    #3 0x7f02df96ff70 in _sceneBuildHelper ../src/loaders/svg/tvgSvgSceneBuilder.cpp:800
    #4 0x7f02df9752dc in svgSceneBuild(SvgLoaderData&, Box, float, float, AspectRatioAlign, AspectRatioMeetOrSlice, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, SvgViewFlag) ../src/loaders/svg/tvgSvgSceneBuilder.cpp:853
    #5 0x7f02df95d652 in SvgLoader::run(unsigned int) ../src/loaders/svg/tvgSvgLoader.cpp:3713
    #6 0x7f02df94a9db in SvgLoader::header() ../src/loaders/svg/tvgSvgLoader.cpp:3796
    #7 0x7f02df95b92b in SvgLoader::open(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ../src/loaders/svg/tvgSvgLoader.cpp:3842
    #8 0x7f02df86bce5 in LoaderMgr::loader(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool*) ../src/renderer/tvgLoader.cpp:307
    #9 0x7f02df88bc5d in tvg::Picture::Impl::load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ../src/renderer/tvgPicture.h:152
    #10 0x7f02df88bc5d in tvg::Picture::load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ../src/renderer/tvgPicture.cpp:175
    #11 0x55e0bfee0c24 in tvgDrawCmds(tvg::Canvas*) ../examples/PictureSvg.cpp:44
    #12 0x55e0bfee1873 in tvgSwTest(unsigned int*) ../examples/PictureSvg.cpp:88
    #13 0x55e0bfedfccd in createSwView ../examples/Common.h:73
    #14 0x55e0bfedfccd in main ../examples/PictureSvg.cpp:158
    #15 0x7f02de004082 in __libc_start_main ../csu/libc-start.c:308
    #16 0x55e0bfedffbd in _start (/home/junsu/dev/os/thorvg/build/examples/PictureSvg+0xbfbd)

0x6030000d631c is located 0 bytes to the right of 28-byte region [0x6030000d6300,0x6030000d631c)
allocated by thread T0 here:
    #0 0x7f02dffa9c3e in __interceptor_realloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:163
    #1 0x7f02df95e2db in tvg::Array<float>::push(float) ../src/common/tvgArray.h:57
    #2 0x7f02df94acc5 in _attrParsePolygonPoints ../src/loaders/svg/tvgSvgLoader.cpp:1728
    #3 0x7f02df94acc5 in _attrParsePolygonNode ../src/loaders/svg/tvgSvgLoader.cpp:1746
    #4 0x7f02df978e45 in simpleXmlParseAttributes(char const*, unsigned int, bool (*)(void*, char const*, char const*), void const*) ../src/loaders/svg/tvgXmlParser.cpp:358
    #5 0x7f02df94df10 in _createPolygonNode ../src/loaders/svg/tvgSvgLoader.cpp:1771
    #6 0x7f02df95823a in _svgLoaderParserXmlOpen ../src/loaders/svg/tvgSvgLoader.cpp:3258
    #7 0x7f02df9589da in _svgLoaderParser ../src/loaders/svg/tvgSvgLoader.cpp:3336
    #8 0x7f02df979006 in simpleXmlParse(char const*, unsigned int, bool, bool (*)(void*, SimpleXMLType, char const*, unsigned int), void const*) ../src/loaders/svg/tvgXmlParser.cpp:433
    #9 0x7f02df95c1ea in SvgLoader::run(unsigned int) ../src/loaders/svg/tvgSvgLoader.cpp:3694
    #10 0x7f02df94a9db in SvgLoader::header() ../src/loaders/svg/tvgSvgLoader.cpp:3796
    #11 0x7f02df95b92b in SvgLoader::open(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ../src/loaders/svg/tvgSvgLoader.cpp:3842
    #12 0x7f02df86bce5 in LoaderMgr::loader(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool*) ../src/renderer/tvgLoader.cpp:307
    #13 0x7f02df88bc5d in tvg::Picture::Impl::load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ../src/renderer/tvgPicture.h:152
    #14 0x7f02df88bc5d in tvg::Picture::load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ../src/renderer/tvgPicture.cpp:175
    #15 0x55e0bfee0c24 in tvgDrawCmds(tvg::Canvas*) ../examples/PictureSvg.cpp:44
    #16 0x55e0bfee1873 in tvgSwTest(unsigned int*) ../examples/PictureSvg.cpp:88
    #17 0x55e0bfedfccd in createSwView ../examples/Common.h:73
    #18 0x55e0bfedfccd in main ../examples/PictureSvg.cpp:158
    #19 0x7f02de004082 in __libc_start_main ../csu/libc-start.c:308

SUMMARY: AddressSanitizer: heap-buffer-overflow ../src/loaders/svg/tvgSvgSceneBuilder.cpp:415 in _recognizeShape
Shadow bytes around the buggy address:
  0x0c0680012c10: fa fa fd fd fd fa fa fa fd fd fd fd fa fa fd fd
  0x0c0680012c20: fd fd fa fa fd fd fd fd fa fa fd fd fd fa fa fa
  0x0c0680012c30: fd fd fd fd fa fa fd fd fd fa fa fa 00 00 00 fa
  0x0c0680012c40: fa fa 00 00 00 fa fa fa 00 00 04 fa fa fa 00 00
  0x0c0680012c50: 00 00 fa fa 00 00 00 fa fa fa 00 00 00 fa fa fa
=>0x0c0680012c60: 00 00 00[04]fa fa 00 00 00 fa fa fa 00 00 00 fa
  0x0c0680012c70: fa fa 00 00 00 00 fa fa fa fa fa fa fa fa fa fa
  0x0c0680012c80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0680012c90: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0680012ca0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0680012cb0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==193713==ABORTING

```

Reply:
gotcha, thanks.